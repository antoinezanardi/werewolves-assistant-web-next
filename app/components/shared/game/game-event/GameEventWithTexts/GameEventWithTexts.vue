<template>
  <div
    id="game-event-with-texts"
    class="flex flex-col h-full"
  >
    <div
      id="game-event-default-slot-container"
      class="grow"
    >
      <slot/>
    </div>

    <GameEventTextsManager
      id="game-event-texts-manager"
      class="grow md:grow-0"
      :texts="props.texts"
      @game-event-text-change="onGameEventTextChangeFromGameEventTextsManager"
    />
  </div>
</template>

<script setup lang="ts">
import type { GameEventWithTextsEmits, GameEventWithTextsProps } from "~/components/shared/game/game-event/GameEventWithTexts/game-event-with-texts.types";
import GameEventTextsManager from "~/components/shared/game/game-event/GameEventWithTexts/GameEventTextsManager/GameEventTextsManager.vue";

const props = defineProps<GameEventWithTextsProps>();

const emit = defineEmits<GameEventWithTextsEmits>();

function onGameEventTextChangeFromGameEventTextsManager(newGameEventText: string | undefined): void {
  if (newGameEventText === undefined) {
    return;
  }
  emit("gameEventTextChange", newGameEventText);
}

defineSlots<{
  default: () => void;
}>();
</script>