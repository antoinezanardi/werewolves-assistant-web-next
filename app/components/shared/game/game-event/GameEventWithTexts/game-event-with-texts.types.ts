type GameEventWithTextsProps = {
  texts: string[];
};

type GameEventWithTextsEmits = {
  gameEventTextChange: [string | undefined];
};

export type {
  GameEventWithTextsProps,
  GameEventWithTextsEmits,
};