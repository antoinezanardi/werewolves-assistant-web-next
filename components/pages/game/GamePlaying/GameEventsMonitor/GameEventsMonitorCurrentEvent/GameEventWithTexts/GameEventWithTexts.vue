<template>
  <div
    id="game-events-with-texts"
    class="flex flex-col"
  >
    <div
      id="game-event-text-manager"
      class="flex"
    >
      <button
        id="previous-event-text-button"
        type="button"
        @click.prevent="previousGameEventText"
      >
        <i class="fa fa-2x fa-chevron-left"/>
      </button>

      <p class="grow">
        {{ currentGameEventText }}
      </p>

      <button
        id="next-event-text-button"
        type="button"
        @click.prevent="nextEventText"
      >
        <i class="fa fa-2x fa-chevron-right"/>
      </button>
    </div>
  </div>
</template>

<script setup lang="ts">
import type { GameEventWithTextsProps } from "~/components/pages/game/GamePlaying/GameEventsMonitor/GameEventsMonitorCurrentEvent/GameEventWithTexts/game-event-with-texts.types";

const props = defineProps<GameEventWithTextsProps>();

const currentIndex = ref<number>(0);

const currentGameEventText = computed<string | undefined>(() => props.texts[currentIndex.value]);

function nextEventText(): void {
  currentIndex.value++;
}

function previousGameEventText(): void {
  currentIndex.value--;
}
</script>