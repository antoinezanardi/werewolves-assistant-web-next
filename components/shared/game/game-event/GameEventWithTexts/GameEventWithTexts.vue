<template>
  <div
    id="game-event-with-texts"
    class="flex flex-col h-full"
  >
    <div
      id="game-event-default-slot-container"
      class="grow"
    >
      <slot/>
    </div>

    <div
      id="game-event-text-manager"
      class="flex gap-4 h-1/2 items-center"
    >
      <button
        id="previous-event-text-button"
        class="ps-4"
        :class="{ 'text-gray-500': !canGoToPreviousGameEventText }"
        :disabled="!canGoToPreviousGameEventText"
        type="button"
        @click.prevent="previousGameEventText"
      >
        <i class="fa fa-3x fa-chevron-left"/>
      </button>

      <div
        id="game-event-text-container"
        class="flex grow h-1/2 items-center justify-center"
      >
        <transition
          mode="out-in"
          name="fade"
        >
          <p
            :key="currentGameEventText"
            class="!mb-0 !text-2xl text-center"
          >
            {{ currentGameEventText }}
          </p>
        </transition>
      </div>

      <button
        id="next-event-text-button"
        class="pe-4"
        type="button"
        @click.prevent="nextEventText"
      >
        <i class="fa fa-3x fa-chevron-right"/>
      </button>
    </div>
  </div>
</template>

<script setup lang="ts">
import type { GameEventWithTextsProps } from "~/components/shared/game/game-event/GameEventWithTexts/game-event-with-texts.types";
import { useGameEventsStore } from "~/stores/game/game-event/useGameEventsStore";

const props = defineProps<GameEventWithTextsProps>();

defineSlots<{
  default: () => void;
}>();

const gameEventsStore = useGameEventsStore();
const { goToNextGameEvent } = gameEventsStore;

const currentIndex = ref<number>(0);

const currentGameEventText = computed<string | undefined>(() => props.texts[currentIndex.value]);

const canGoToPreviousGameEventText = computed<boolean>(() => currentIndex.value > 0);

function nextEventText(): void {
  currentIndex.value++;
  if (currentGameEventText.value === undefined) {
    goToNextGameEvent();
  }
}

function previousGameEventText(): void {
  currentIndex.value--;
}
</script>