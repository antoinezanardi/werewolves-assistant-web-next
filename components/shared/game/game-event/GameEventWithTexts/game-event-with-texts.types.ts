type GameEventWithTextsProps = {
  texts: string[];
};

export type { GameEventWithTextsProps };