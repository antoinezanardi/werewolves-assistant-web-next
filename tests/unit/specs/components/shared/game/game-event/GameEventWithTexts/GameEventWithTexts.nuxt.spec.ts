import type { mount } from "@vue/test-utils";
import type { ComponentMountingOptions } from "@vue/test-utils/dist/mount";
import type { GameEventWithTextsProps } from "~/components/shared/game/game-event/GameEventWithTexts/game-event-with-texts.types";
import type GameEventTextsManager from "~/components/shared/game/game-event/GameEventWithTexts/GameEventTextsManager/GameEventTextsManager.vue";
import GameEventWithTexts from "~/components/shared/game/game-event/GameEventWithTexts/GameEventWithTexts.vue";

import { mountSuspendedComponent } from "@tests/unit/utils/helpers/mount.helpers";

describe("Game Event With Texts Component", () => {
  let wrapper: ReturnType<typeof mount<typeof GameEventWithTexts>>;
  const defaultProps: GameEventWithTextsProps = {
    texts: [
      "Game starts.",
      "Day rises.",
      "Night falls.",
    ],
  };

  async function mountGameEventWithTextsComponent(options: ComponentMountingOptions<typeof GameEventWithTexts> = {}):
  Promise<ReturnType<typeof mount<typeof GameEventWithTexts>>> {
    return mountSuspendedComponent(GameEventWithTexts, {
      props: defaultProps,
      ...options,
    });
  }

  beforeEach(async() => {
    wrapper = await mountGameEventWithTextsComponent();
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should match snapshot without shallow when rendered.", async() => {
    wrapper = await mountGameEventWithTextsComponent({ shallow: false });

    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Game Event Texts Manager", () => {
    it("should pass texts from props when rendered.", () => {
      const gameEventTextsManager = wrapper.findComponent<typeof GameEventTextsManager>("#game-event-texts-manager");

      expect(gameEventTextsManager.props("texts")).toStrictEqual<string[]>(defaultProps.texts);
    });
  });
});