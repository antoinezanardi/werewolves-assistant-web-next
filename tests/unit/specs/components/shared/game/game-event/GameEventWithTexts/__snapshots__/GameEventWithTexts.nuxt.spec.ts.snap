// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game Event With Texts Component > should match snapshot when rendered. 1`] = `
"<div id="game-event-with-texts" class="flex flex-col h-full">
  <div id="game-event-default-slot-container" class="grow"></div>
  <game-event-texts-manager-stub texts="Game starts.,Day rises.,Night falls." id="game-event-texts-manager"></game-event-texts-manager-stub>
</div>"
`;

exports[`Game Event With Texts Component > should match snapshot without shallow when rendered. 1`] = `
"<div id="game-event-with-texts" class="flex flex-col h-full">
  <div id="game-event-default-slot-container" class="grow"></div>
  <div id="game-event-texts-manager" class="flex gap-4 h-1/2 items-center"><button id="game-event-previous-text-button" aria-label="Back to the previous event text" class="ps-4 text-gray-500" disabled="" type="button" data-pd-tooltip="true"><i class="fa fa-3x fa-chevron-left"></i></button>
    <div id="game-event-text-container" class="flex grow h-1/2 items-center justify-center">
      <transition-stub mode="out-in" name="fade" appear="false" persisted="false" css="true">
        <p id="current-event-text" class="!mb-0 !text-2xl text-center">Game starts.</p>
      </transition-stub>
    </div><button id="game-event-next-text-button" aria-label="Next event text" class="pe-4" type="button" data-pd-tooltip="true"><i class="fa fa-3x fa-chevron-right"></i></button>
  </div>
</div>"
`;
